shared/src/types/ibc.rs

```diff
@@ -600,3 +601,549 @@ impl ConnectionOpenConfirmData {
         .map_err(Error::DecodingError)
     }
 }
+
+/// Data to initialize a channel
+#[derive(Debug, Clone, BorshSerialize, BorshDeserialize)]
+pub struct ChannelOpenInitData {
+    /// The port ID
+    port_id: String,
+    /// The order feature of the channel
+    order: String,
+    /// The counterparty
+    counterparty: Vec<u8>,
+    /// The connection hops
+    connection_hops: Vec<String>,
+    /// The version
+    version: String,
+}
+
+impl ChannelOpenInitData {
+    /// Returns the data to initalize a channel
+    pub fn new(
+        port_id: PortId,
+        order: Order,
+        counterparty: ChanCounterparty,
+        connection_hops: Vec<ConnectionId>,
+        version: String,
+    ) -> Self {
+        let port_id = port_id.as_str().to_owned();
+        let order = order.as_str().to_owned();
+        let counterparty = counterparty
+            .encode_vec()
+            .expect("Encoding a counterparty shouldn't fail");
+        let connection_hops = connection_hops
+            .iter()
+            .map(|c| c.as_str().to_owned())
+            .collect();
+        Self {
+            port_id,
+            order,
+            counterparty,
+            connection_hops,
+            version,
+        }
+    }
+
+    /// Returns the port ID
+    pub fn port_id(&self) -> Result<PortId> {
+        PortId::from_str(&self.port_id)
+            .map_err(|e| Error::DecodingError(e.to_string()))
+    }
+
+    /// Returns the order
+    pub fn order(&self) -> Result<Order> {
+        Order::from_str(&self.order)
+            .map_err(|e| Error::DecodingError(e.to_string()))
+    }
+
+    /// Returns the counterparty
+    pub fn counterparty(&self) -> Result<ChanCounterparty> {
+        ChanCounterparty::decode_vec(&self.counterparty)
+            .map_err(|e| Error::DecodingError(e.to_string()))
+    }
+
+    /// Returns the connection hops
+    pub fn connection_hops(&self) -> Result<Vec<ConnectionId>> {
+        let mut hops = vec![];
+        for conn_str in &self.connection_hops {
+            let id = ConnectionId::from_str(conn_str)
+                .map_err(|e| Error::DecodingError(e.to_string()))?;
+            hops.push(id);
+        }
+        Ok(hops)
```

Comment:
You can do this, because `Result` implements `FromIter`:
```suggestion
        self
            .connection_hops
            .iter()
            .map(|conn_str| {
                ConnectionId::from_str(conn_str)
                    .map_err(|e| Error::DecodingError(e.to_string()))
            })
            .collect()
```
But unfortunately it's actually longer than the first version :(

Alternatively, with the first version it would be better to change the first line to pre-allocate the vec for the needed size `let mut hops = Vec::with_capacity(self.connection_hops.len())`.

Reply:
Thanks! I've modified it to use the iteration!